---
title: "CHIP: internal data characterization"
author: "Lauren Fairchild lauren.fairchild@novartis.com"
output: html_document
---

Companion code for "Clonal hematopoiesis detection in cancer patients using cell free DNA sequencing" manuscript. This script uses annotated published SNV data from MSKCC (Razavi et al 2019) to train logistic regression and random forest models to predict the biological source of SNVs in blood plasma. The following elements are generated by this script:

* Figure 2c
* Figure S3
* Figure 3a,b,c
* Figure S4a,b
* Figure S5

```{r load libraries, warning=FALSE, message=FALSE}
print(date())

library(ggsci)
library(cowplot)
library(ggplot2)
library(RColorBrewer)
set.seed(123)

pat_data = read.csv("../input/Data_file_S2_patient_info.csv", header=T)
anon_snv_calls = read.csv("../input/Data_file_S3_SNV_calls.csv")
anon_indel_calls = read.csv("../input/Data_file_S4_INDEL_calls.csv")

```

Figure 2c: CH prediction by age line graph
```{r}
cols = pal_npg()(8)
plot_cols = c(CH_canonical=cols[1], CH_myeloid=cols[2], ND="grey25",NO_CH=cols[3])

freqs = aggregate(pat_data$CHIP_STATUS, by=list(pat_data$age_bin), function(x){table(x) / length(x)})
freqs$counts = as.vector(table(pat_data$age_bin))

for(i in 1:ncol(freqs$x)){
  if(i == 1){
    plot(freqs$x[c(2:6,1),i], type='l', ylim=c(0,0.6), col=plot_cols[colnames(freqs$x)[i]], lwd=2, xaxt="n", xlab="Age", ylab="Proportion of patients")
    axis(1, 1:length(freqs$Group.1), paste0(freqs$Group.1[c(2:6,1)], "\nN=", freqs$counts[c(2:6,1)]), tick=F)
  }else{
    lines(freqs$x[c(2:6,1),i], col=plot_cols[colnames(freqs$x)[i]], lwd=2)
  }
}
legend("topleft", names(plot_cols), fill=plot_cols, bty='n')

# p-value for difference in fraction of CH+ patients in oldest and youngest groups (fisher's exact test)
fisher.mat = matrix(c(nrow(subset(pat_data, age_bin %in% c("20-29","30-39") & CHIP_STATUS == "CH_canonical")),
                      nrow(subset(pat_data, age_bin %in% c("20-29","30-39") & CHIP_STATUS != "CH_canonical")),
                      nrow(subset(pat_data, age_bin %in% c(">70") & CHIP_STATUS == "CH_canonical")),
                      nrow(subset(pat_data, age_bin %in% c(">70") & CHIP_STATUS != "CH_canonical"))), nrow=2)
print (fisher.mat)
print(fisher.test(fisher.mat))

# repeat for PDF output
pdf("../output/Figure_2c.pdf", width=7, height=6, useDingbats = F)
for(i in 1:ncol(freqs$x)){
  if(i == 1){
    plot(freqs$x[c(2:6,1),i], type='l', ylim=c(0,0.6), col=plot_cols[colnames(freqs$x)[i]], lwd=2, xaxt="n", xlab="Age", ylab="Proportion of patients")
    axis(1, 1:length(freqs$Group.1), paste0(freqs$Group.1[c(2:6,1)], "\nN=", freqs$counts[c(2:6,1)]), tick=F)
  }else{
    lines(freqs$x[c(2:6,1),i], col=plot_cols[colnames(freqs$x)[i]], lwd=2)
  }
}
legend("topleft", names(plot_cols), fill=plot_cols, bty='n')
dev.off()
```

Figure S3 - Comparison of CH incidence in cancer and healthy controls (from other published works)
```{r}
pat_subset = subset(pat_data, !is.na(age_bin))

freqs_noAF = aggregate(pat_subset$CHIP_STATUS, by=list(pat_subset$age_bin), function(x){table(x) / length(x)})
freqs_noAF = data.frame(age.bin=freqs_noAF$Group.1, freqs_noAF$x)

pat_subset$CHIP_STATUS_2per = "NO_CH"
pat_subset$CHIP_STATUS_2per[!is.na(pat_subset$max.obs.myeloid.chip.af) & pat_subset$max.obs.myeloid.chip.af > 0.02] = "CH_myeloid"
pat_subset$CHIP_STATUS_2per[!is.na(pat_subset$max.obs.canon.chip.af) & pat_subset$max.obs.canon.chip.af > 0.02] = "CH_canonical"
freqs_AF = aggregate(pat_subset$CHIP_STATUS_2per, by=list(pat_subset$age_bin), function(x){table(x) / length(x)})
freqs_AF = data.frame(age.bin=freqs_AF$Group.1, freqs_AF$x)

# Bolton et al, 2020 (extended data table 1)
bolton_freq = data.frame(age.bin=c("0-9","10-19","20-29","30-39","40-49","50-59","60-69","70-79","80-89"),
                         CH=c(0.039, 0.044, 0.051, 0.08, 0.13, 0.22, 0.36, 0.50, 0.63))


# Jaiswal et al, 2014 (calculated from Figure 1)
jaiswal_freq = data.frame(age.bin=c("20-29","30-39","40-49","50-59","60-69","70-79","80-89","90-99","100-108"),
                          CH=c(0, 0.001, 0.017, 0.025, 0.056, 0.095, 0.117, 0.163, 0.294))

all_freqs = data.frame(age.bin=c("0-9","10-19","20-29","30-39","40-49","50-59","60-69","70-79","80-89","90-99","100-108"))
all_freqs$Bolton = bolton_freq[match(all_freqs$age.bin, bolton_freq$age.bin), "CH"]
all_freqs$CHpos_CHmye = rowSums(freqs_noAF[match(all_freqs$age.bin, freqs_noAF$age.bin), c("CH_canonical","CH_myeloid")])
all_freqs$CHpos = freqs_noAF[match(all_freqs$age.bin, freqs_noAF$age.bin), "CH_canonical"]
all_freqs$Jaiswal_2014 = jaiswal_freq[match(all_freqs$age.bin, jaiswal_freq$age.bin), "CH"]
all_freqs$CHpos_CHmye_2p = rowSums(freqs_AF[match(all_freqs$age.bin, freqs_AF$age.bin), c("CH_canonical","CH_myeloid")])
all_freqs$CHmye = freqs_noAF[match(all_freqs$age.bin, freqs_noAF$age.bin), "CH_myeloid"]
all_freqs$CHpos_2p = freqs_AF[match(all_freqs$age.bin, freqs_AF$age.bin), "CH_canonical"]
all_freqs$CHmye_2p = freqs_AF[match(all_freqs$age.bin, freqs_AF$age.bin), "CH_myeloid"]


for(i in 2:ncol(all_freqs)){
  if(i == 2){
    plot(1:nrow(all_freqs), all_freqs[,i], col=cols[i-1], type='l', lwd=2, xaxt="n", ylim=c(0,0.7),
         xlab="Age", ylab="Proportion of positive patients")
    axis(1, at=1:nrow(all_freqs), labels=all_freqs[,1], cex=0.8)
    legend("topleft", colnames(all_freqs)[2:ncol(all_freqs)], fill=cols, bty='n')
  }else{
    lines(1:nrow(all_freqs), all_freqs[,i], col=cols[i-1], type='l', lwd=2)
  }
}

# repeat for PDF output
pdf("../output/Figure_S3.pdf", width=10, height=7, useDingbats = F)
for(i in 2:ncol(all_freqs)){
  if(i == 2){
    plot(1:nrow(all_freqs), all_freqs[,i], col=cols[i-1], type='l', lwd=2, xaxt="n", ylim=c(0,0.7),
         xlab="Age", ylab="Proportion of positive patients")
    axis(1, at=1:nrow(all_freqs), labels=all_freqs[,1], cex=0.8)
    legend("topleft", colnames(all_freqs)[2:ncol(all_freqs)], fill=cols, bty='n')
  }else{
    lines(1:nrow(all_freqs), all_freqs[,i], col=cols[i-1], type='l', lwd=2)
  }
}
dev.off()
```


Figure 3A - CH patients split by indication + age distribution panel

Note that the age distribution panel is slightly different from the published version as the patient ages needed to be grouped into 10-year bins for anonymity.
```{r, fig.width = 10, fig.height = 5}

pat_subset = subset(pat_data, !is.na(age_bin) & !is.na(INDIC))
pat_subset$age_random = as.numeric(gsub(">", "", gsub("-.*", "", pat_subset$age_bin))) + floor(runif(nrow(pat_subset),min=0,max=10)) + 1

pat_subset$CH = as.numeric(pat_subset$CHIP_STATUS == "CH_canonical")

# Use only indics with N > 100
pat_subset_indics = subset(pat_subset, unlist(INDIC) %in% c("NON-SMALL CELL LUNG CANCER", "TRIPLE NEGATIVE BREAST CANCER", "BREAST CANCER","CUTANEOUS MELANOMA","COLORECTAL CANCER"))


ch_indic_means = aggregate(pat_subset_indics$CH, by=list(pat_subset_indics$INDIC), mean)
ch_indic_means$x = ch_indic_means$x*100
barplot = ggplot(ch_indic_means, aes(x=reorder(Group.1, x), y=x)) + geom_bar(stat="identity") + ylim(0,100) + 
  ylab("Percent of patients CH+") + xlab("Indication") + coord_flip() + theme_bw()


pat_subset_indics$INDIC = factor(pat_subset_indics$INDIC, levels=c("NON-SMALL CELL LUNG CANCER","CUTANEOUS MELANOMA","BREAST CANCER","COLORECTAL CANCER","TRIPLE NEGATIVE BREAST CANCER"))
density_plot = ggplot(pat_subset_indics, aes(x=age_random)) + geom_density() + xlim(19,91) +
    theme_bw() + theme(strip.background = element_blank()) + facet_wrap(facets=vars(pat_subset_indics$INDIC), ncol=1)


plot_grid(barplot, density_plot, nrow=1, rel_widths = c(3,1))

pdf("../output/Figure_3a.pdf", width=15, height=7, useDingbats = F)
plot_grid(barplot, density_plot, nrow=1, rel_widths = c(3,1))
dev.off()

```

Figure 3B - CH odds ratio by indication, accounting for age

Note that this output is slightly different from the published version as the patient ages needed to be grouped into 10-year bins for anonymity.
```{r}
# Select patients with known age and indication, compare only CH-positive vs CH-negative (remove CH-myeloid)
pat_subset = subset(pat_data, !is.na(age_bin) & !is.na(INDIC))
pat_subset$age = as.numeric(gsub(">", "", gsub("-.*", "", pat_subset$age_bin)))
pat_subset$CH = NA
pat_subset$CH[pat_subset$CHIP_STATUS == "CH_canonical"] = 1
pat_subset$CH[pat_subset$CHIP_STATUS %in% c("ND","NO_CH")] = 0
pat_subset = subset(pat_subset, !is.na(CH))


# Use only indics with N > 100
pat_subset_indics = subset(pat_subset, unlist(INDIC) %in% c("NON-SMALL CELL LUNG CANCER", "TRIPLE NEGATIVE BREAST CANCER", "BREAST CANCER","CUTANEOUS MELANOMA","COLORECTAL CANCER"))

model = glm(pat_subset_indics$CH ~ unlist(pat_subset_indics$INDIC) + pat_subset_indics$age, family="binomial")
summary(model)


plot_odds = function(model, num_indics, ...){
  num_lines = num_indics - 1
  
  # Get confidence interval values
  confint = as.data.frame(exp(confint(model)))
  confint$estimate = exp(coef(summary(model))[,"Estimate"])
  confint = confint[2:num_indics,] 
  
  # Add pvalue to confint table
  confint$pvalue = coef(summary(model))[2:num_indics,"Pr(>|z|)"]
  confint = confint[order(confint$estimate),]
  
  
  # Plot logistic regression estimates with 95% confidence intervals
  par(mar=c(5,17,3,2))
  plot(confint$estimate, 1:num_lines, xlim=c(min(confint$`2.5 %`), max(confint$`97.5 %`)), yaxt='n', xlab="Odds Ratio", ylab="", pch=16, ylim=c(1,4.2), ...)
  axis(2, at=1:num_lines, labels=gsub(".*)", "", rownames(confint)), las=2)
  segments(confint$`2.5 %`, 1:num_lines, confint$`97.5 %`, 1:num_lines)
  abline(v=1, lty=2)
  # Add * for significance
  sig.points = c("***","**","*","")[.bincode(confint$pvalue, breaks=c(0,0.001,0.01,0.05,1))]
  text(confint$estimate, (1:num_lines)+0.15, sig.points, cex=1.2)

}

plot_odds(model, num_indics=5)

pdf("../output/Figure_3b.pdf", width=7, height=4, useDingbats = F)
plot_odds(model, num_indics=5)
dev.off()

```

Figure S4a - CH odds ratio correcting for age and therapy

Note that this output is slightly different from the published version as the patient ages needed to be grouped into 10-year bins for anonymity.
```{r}
# Select patients with known age, indication, and therapy history, compare only CH-positive vs CH-negative (remove CH-myeloid)
pat_subset = subset(pat_data, !is.na(age_bin) & !is.na(INDIC) & !is.na(chemo_last_line) & !is.na(chemo_any_line))
pat_subset$age = as.numeric(gsub(">", "", gsub("-.*", "", pat_subset$age_bin)))
pat_subset$CH = NA
pat_subset$CH[pat_subset$CHIP_STATUS == "CH_canonical"] = 1
pat_subset$CH[pat_subset$CHIP_STATUS %in% c("ND","NO_CH")] = 0
pat_subset = subset(pat_subset, !is.na(CH))

# Use only indics with N > 100
pat_subset_indics = subset(pat_subset, unlist(INDIC) %in% c("NON-SMALL CELL LUNG CANCER", "TRIPLE NEGATIVE BREAST CANCER", "BREAST CANCER","CUTANEOUS MELANOMA","COLORECTAL CANCER"))

# Panel 1: correcting only for age for patients with therapy information curated
model_panel1 = glm(pat_subset_indics$CH ~ unlist(pat_subset_indics$INDIC) + pat_subset_indics$age, family="binomial")
plot_odds(model_panel1, num_indics=5)

# Panel 2: correcting for age and chemo as the most recent line of therapy 
model_panel2 = glm(pat_subset_indics$CH ~ unlist(pat_subset_indics$INDIC) + pat_subset_indics$age + pat_subset_indics$chemo_last_line, family="binomial")
plot_odds(model_panel2, num_indics = 5)

# Panel 3: correcting for age and chemo as any line of therapy 
model_panel3 = glm(pat_subset_indics$CH ~ unlist(pat_subset_indics$INDIC) + pat_subset_indics$age + pat_subset_indics$chemo_any_line, family="binomial")
plot_odds(model_panel3, num_indics = 5)


pdf("../output/Figure_S4a.pdf", width=15, height=3, useDingbats = F)
par(mfrow=c(1,3))
plot_odds(model_panel1, num_indics=5)
plot_odds(model_panel2, num_indics=5)
plot_odds(model_panel3, num_indics=5)
dev.off()

```

Figure S4b - CH/CH-myeloid odds ratio by indication, accounting for age
```{r}
# Select patients with known age and indication, compare CH-positive OR CH-myeloid vs CH-negative
pat_subset = subset(pat_data, !is.na(age_bin) & !is.na(INDIC))
pat_subset$age = as.numeric(gsub(">", "", gsub("-.*", "", pat_subset$age_bin)))
pat_subset$CH = NA
pat_subset$CH[pat_subset$CHIP_STATUS %in% c("CH_canonical", "CH_myeloid")] = 1
pat_subset$CH[pat_subset$CHIP_STATUS %in% c("ND","NO_CH")] = 0
pat_subset = subset(pat_subset, !is.na(CH))

# Use only indics with N > 100
pat_subset_indics = subset(pat_subset, unlist(INDIC) %in% c("NON-SMALL CELL LUNG CANCER", "TRIPLE NEGATIVE BREAST CANCER", "BREAST CANCER","CUTANEOUS MELANOMA","COLORECTAL CANCER"))

model = glm(pat_subset_indics$CH ~ unlist(pat_subset_indics$INDIC) + pat_subset_indics$age, family="binomial")

plot_odds(model, num_indics=5)

pdf("../output/Figure_S4b.pdf", width=7, height=4, useDingbats = F)
plot_odds(model, num_indics=5)
dev.off()
```

Figure S5 - Odds ratios for specific genes
```{r}
genes_of_interest = c("DNMT3A","TET2","ASXL1","JAK2","PPM1D","SF3B1")#,"TP53")
indics = c("NON-SMALL CELL LUNG CANCER", "TRIPLE NEGATIVE BREAST CANCER", "BREAST CANCER","CUTANEOUS MELANOMA","COLORECTAL CANCER")

snv_calls = read.csv("/mnt/tmplabdata/ngdx/projects/methods_dev/CHIP/faircla1/manuscript_code_data/input/Anon_SNV_calls.csv", stringsAsFactors = F)
indel_calls = read.csv("/mnt/tmplabdata/ngdx/projects/methods_dev/CHIP/faircla1/manuscript_code_data/input/Anon_INDEL_calls.csv", stringsAsFactors = F)

for(gene in genes_of_interest){
  pos_pats = unique(c(subset(snv_calls, gene_name == gene)$anon_ID,
                      subset(indel_calls, gene_name == gene)$anon_ID))

  pat_subset = subset(pat_data, !is.na(age_bin) & !is.na(INDIC))
  pat_subset$age = as.numeric(gsub(">", "", gsub("-.*", "", pat_subset$age_bin)))
  pat_subset$CH = c(0,1)[as.numeric(pat_subset$anon_ID %in% pos_pats)+1]
  pat_subset_indics = subset(pat_subset, unlist(INDIC) %in% indics)
  
  # Remove indications with 0 mutations in gene of interest
  indics_keep = indics[indics %in% subset(pat_subset_indics, INDIC %in% indics & CH==1)$INDIC]
  pat_subset_indics = subset(pat_subset_indics, INDIC %in% indics_keep)
  
  model = glm(pat_subset_indics$CH ~ unlist(pat_subset_indics$INDIC) + pat_subset_indics$age, family="binomial")

  plot_odds(model, length(indics_keep), main=gene)
}


# repeat for PDF output
pdf("../output/Figure_S5.pdf", width=20, height=7, useDingbats = F)
par(mfrow=c(2,4))
for(gene in genes_of_interest){
  pos_pats = unique(c(subset(snv_calls, gene_name == gene)$anon_ID,
                      subset(indel_calls, gene_name == gene)$anon_ID))

  pat_subset = subset(pat_data, !is.na(age_bin) & !is.na(INDIC))
  pat_subset$age = as.numeric(gsub(">", "", gsub("-.*", "", pat_subset$age_bin)))
  pat_subset$CH = c(0,1)[as.numeric(pat_subset$anon_ID %in% pos_pats)+1]
  pat_subset_indics = subset(pat_subset, unlist(INDIC) %in% indics)
  
  # Remove indications with 0 mutations in gene of interest
  indics_keep = indics[indics %in% subset(pat_subset_indics, INDIC %in% indics & CH==1)$INDIC]
  pat_subset_indics = subset(pat_subset_indics, INDIC %in% indics_keep)
  
  model = glm(pat_subset_indics$CH ~ unlist(pat_subset_indics$INDIC) + pat_subset_indics$age, family="binomial")

  plot_odds(model, length(indics_keep), main=gene)
}

dev.off()

```

Figure 3C - Barplots of CH gene frequency by indication 
```{r}
# create matrix of counts, CH gene + by indication

ch_freqs = do.call(cbind, lapply(genes_of_interest, function(gene){
  pos_pats = unique(c(subset(snv_calls, gene_name == gene)$anon_ID,
                      subset(indel_calls, gene_name == gene)$anon_ID))
  pat_subset = subset(pat_data, !is.na(age_bin) & !is.na(INDIC))
  pat_subset$age = as.numeric(gsub(">", "", gsub("-.*", "", pat_subset$age_bin)))
  pat_subset$CH = c(0,1)[as.numeric(pat_subset$anon_ID %in% pos_pats)+1]
  pat_subset_indics = subset(pat_subset, unlist(INDIC) %in% indics)
  
  
  # proportion of patients with variant in gene
  counts = table(pat_subset_indics$INDIC)[indics]
  s = subset(pat_subset_indics, CH == 1)
  table(s$INDIC)[indics] / counts
}))

colnames(ch_freqs) = genes_of_interest

colors = brewer.pal(5, "Set2")
names(colors) = indics

barplot(ch_freqs, beside=T, col=colors, ylab="Proportion of patients with variant")
legend("top", indics, fill=colors, bty='n', cex=0.8)

pdf("../output/Figure_3c.pdf", width=7, height=4, useDingbats = F)
barplot(ch_freqs, beside=T, col=colors, ylab="Proportion of patients with variant")
legend("top", indics, fill=colors, bty='n', cex=0.8)
dev.off()
```


```{r}
sessionInfo()
```

